internal/provider/resource_monitor_resource.go

```diff
@@ -0,0 +1,807 @@
+	resp.TypeName = req.ProviderTypeName + "_resource_monitor"
+}
+
+func (r *ResourceMonitorResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = resourceMonitorSchemaV1()
+}
+
+func (r *ResourceMonitorResource) UpgradeState(ctx context.Context) map[int64]resource.StateUpgrader {
+	schemaV0 := resourceMonitorSchemaV0()
+	return map[int64]resource.StateUpgrader{
+		// State upgrade implementation from 0 to 1
+		0: {
+			PriorSchema:   &schemaV0,
+			StateUpgrader: upgradeResourceMonitorStateV0toV1,
+		},
+	}
+}
+
+func (r *ResourceMonitorResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	providerData, ok := req.ProviderData.(*ProviderData)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *sdk.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = providerData.client
+}
+
+func (r *ResourceMonitorResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	// we aren't really modifying the plan, just logging what the plan intends to do
+	resp.Plan = req.Plan
+	var plan, state *resourceMonitorModelV1
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resourceName := "snowflake_database"
+	// DELETE
+	if req.Plan.Raw.IsNull() {
+		_, readLogs, _ := r.read(ctx, state, true)
+		_, deleteLogs, _ := r.delete(ctx, state, true)
+		deleteLogs = append(deleteLogs, readLogs...)
+		tflog.Debug(ctx, formatSQLPreview(DeleteOperation, resourceName, state.Id.ValueString(), deleteLogs))
+		return
+	}
+
+	// CREATE
+	if plan.Id.IsUnknown() {
+		_, createLogs, _ := r.create(ctx, plan, true)
+		plan.Id = types.StringValue(sdk.NewAccountObjectIdentifier(plan.Name.ValueString()).FullyQualifiedName())
+		_, readLogs, _ := r.read(ctx, plan, true)
+		createLogs = append(createLogs, readLogs...)
+		tflog.Debug(ctx, formatSQLPreview(CreateOperation, resourceName, "", createLogs))
+		return
+	}
+
+	if plan.Equals(state, ctx) {
+		// READ
+		_, logs, _ := r.read(ctx, state, true)
+		tflog.Debug(ctx, formatSQLPreview(ReadOperation, resourceName, state.Id.ValueString(), logs))
+		return
+	} else {
+		// UPDATE
+		_, updateLogs, _ := r.update(ctx, plan, state, true)
+		_, readLogs, _ := r.read(ctx, plan, true)
+		updateLogs = append(updateLogs, readLogs...)
+		tflog.Debug(ctx, formatSQLPreview(UpdateOperation, resourceName, state.Id.ValueString(), updateLogs))
+	}
+}
+
+func (r *ResourceMonitorResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
```

Comment:
I am not commenting much below, because it seems like logic similar to the one in the existing provider, I see that there are some additions and shortcuts but we will probably talk about them on Monday.